prow/cmd/pipeline/controller.go

```diff
@@ -412,7 +432,25 @@ func reconcile(c reconciler, key string) error {
 		case !ok, v != "true":
 			return nil
 		}
+
+		logrus.Infof("Delete PipelineResources for PipelineRun/%s", key)
+
+		for _, resourceBinding := range p.Spec.Resources {
+			res, err := c.getPipelineResource(ctx, namespace, resourceBinding.ResourceRef.Name)
+
+			if err != nil {
+				return fmt.Errorf("delete pipelineresource: %v", err)
+			}
+
+			if label, ok := res.Labels[kube.CreatedByProw]; ok && label == "true" {
+				if err = c.deletePipelineResource(ctx, namespace, res.Name); err != nil {
```

Comment:
IIUC, this will only work if we still have a ProwJob, right? Would it make sense to add a finalizer to the prowjob when we create the pipelineResource to make sure it stays until we cleaned up?

Reply:
The pipelinerun is deleted when the prowjob is no more, as long as the prowjob exists the pipelinerun is not deleted.
The loop adds deletion of pipelineresources created by prow that belong to the pipelinerun being deleted, but when it happens the prowjob is already gone.

Reply:
Okay, that moves the responsibility and the finalizer down to the pipelinerun, but the point per se remains: We must use a finalizer, as Kube doesn't give us a guarantee that we observe certain events

Reply:
I see your point, i could take some time to switch to finalizers but imho it should be another pull request.
This one is about improving resources cleanup, i will make another one making the detection process more robust.

Reply:
Maybe I am missing something, but adding a finalizer to the pipelinerun before we create the pipelineresource and removing it when we delete the pipelineresource/get a 404 for the pipelineresource shouldn't be very difficult?

This assumes that we always delete the pipelineresource before we delete the pipelinerun thought.

Reply:
I have never used finalizers before.
I will read about it and try to use that instead if you prefer.